source/pkcs11.c

```diff
@@ -12,3 +20,359 @@
+        case (CKR_PUBLIC_KEY_INVALID): return "CKR_PUBLIC_KEY_INVALID";
+        case (CKR_FUNCTION_REJECTED): return "CKR_FUNCTION_REJECTED";
+        case (CKR_VENDOR_DEFINED): return "CKR_VENDOR_DEFINED";
+        default: return "<UNKNOWN ERROR CODE>";
+    }
+    /* clang-format on */
+}
+
+/* Log the failure of a PKCS#11 function, and call aws_raise_error() with the appropriate AWS error code */
+static int s_raise_ck_error(const struct aws_pkcs11_lib *pkcs11_lib, const char *fn_name, CK_RV rv) {
+    /* For now, we just have one AWS error code for all PKCS#11 errors */
+    int aws_err = AWS_IO_PKCS11_ERROR;
+
+    AWS_LOGF_ERROR(
+        AWS_LS_IO_PKCS11,
+        "id=%p: %s() failed. PKCS#11 error: %s (0x%08lX). AWS error: %s.",
+        (void *)pkcs11_lib,
+        fn_name,
+        s_ckr_str(rv),
+        rv,
+        aws_error_name(aws_err));
+
+    return aws_raise_error(aws_err);
+}
+
+/* PKCS#11 often pads strings with ' ' */
+static bool s_is_padding(uint8_t c) {
+    return c == ' ';
+}
+
+/* Return byte-cursor to string with ' ' padding trimmed off.
+ * PKCS#11 structs commonly stores strings in fixed-width arrays, padded by ' ' instead of null-terminator */
+static struct aws_byte_cursor s_trim_padding(const uint8_t *str, size_t len) {
+    const struct aws_byte_cursor src = aws_byte_cursor_from_array(str, len);
+    return aws_byte_cursor_right_trim_pred(&src, s_is_padding);
+}
+
+/* Callback for PKCS#11 library to create a mutex.
+ * Described in PKCS11-base-v2.40 section 3.7 */
+static CK_RV s_pkcs11_create_mutex(CK_VOID_PTR_PTR mutex_out) {
+    if (mutex_out == NULL) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    /* Using the default allocator because there's no way to know which PKCS#11 instance is invoking this callback */
+    struct aws_allocator *allocator = aws_default_allocator();
+
+    struct aws_mutex *mutex = aws_mem_calloc(allocator, 1, sizeof(struct aws_mutex));
+    if (aws_mutex_init(mutex)) {
+        AWS_LOGF_ERROR(AWS_LS_IO_PKCS11, "PKCS#11 CreateMutex() failed, error %s", aws_error_name(aws_last_error()));
+        aws_mem_release(allocator, mutex);
+        *mutex_out = NULL;
+        return CKR_GENERAL_ERROR;
+    }
+
+    *mutex_out = mutex;
+    return CKR_OK;
+}
+
+/* Callback for PKCS#11 library to destroy a mutex.
+ * Described in PKCS11-base-v2.40 section 3.7 */
+static CK_RV s_pkcs11_destroy_mutex(CK_VOID_PTR mutex_ptr) {
+    if (mutex_ptr == NULL) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    struct aws_mutex *mutex = mutex_ptr;
+    aws_mutex_clean_up(mutex);
+    aws_mem_release(aws_default_allocator(), mutex);
+    return CKR_OK;
+}
+
+/* Callback for PKCS#11 library to lock a mutex.
+ * Described in PKCS11-base-v2.40 section 3.7 */
+static CK_RV s_pkcs11_lock_mutex(CK_VOID_PTR mutex_ptr) {
+    if (mutex_ptr == NULL) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    struct aws_mutex *mutex = mutex_ptr;
+    if (aws_mutex_lock(mutex)) {
```

Comment:
What is the behavior when this mutex has been already acquired? Do we block or fail?

Reply:
It just calls through to [pthread_mutex_lock()](https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html#ERRORS) on Unix or [AcquireSRWLockExclusive()](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-acquiresrwlockexclusive) on Windows, so it's going to do whatever they would do.

So if the mutex has been acquired on another thread, it will block.
If it has already been acquired on this thread it will deadlock.